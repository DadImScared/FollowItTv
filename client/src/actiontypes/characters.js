
export const ADD_CHARACTERS = 'characters/ADD_CHARACTERS';
