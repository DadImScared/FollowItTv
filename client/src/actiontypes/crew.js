
export const ADD_CREW = 'crew/ADD_CREW';
