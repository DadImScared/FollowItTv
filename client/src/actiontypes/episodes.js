
export const ADD_EPISODES = 'episodes/ADD_EPISODES';
