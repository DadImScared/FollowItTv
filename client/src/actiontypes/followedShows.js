
export const FOLLOW_SHOW = 'followedShows/FOLLOW_SHOW';
export const UNFOLLOW_SHOW = 'followedShows/UNFOLLOW_SHOW';
export const ADD_DAY = 'followedShows/ADD_DAY';
