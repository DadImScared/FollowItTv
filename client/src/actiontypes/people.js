
export const ADD_PEOPLE = 'people/ADD_PEOPLE';
