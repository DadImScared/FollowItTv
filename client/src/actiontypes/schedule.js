
export const ADD_SCHEDULE = 'schedule/ADD_SCHEDULE';
