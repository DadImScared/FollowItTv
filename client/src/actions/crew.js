
import axios from 'axios';
import _ from 'lodash';
import { normalize } from 'normalizr';

import { ADD_CREW } from '../actiontypes/crew';
import { loadingStart, loadingEnd } from './loading';
import { baseUrl } from './seasons';
import { crewSchema } from '../schemas/person';

export const addCrew = (entities) => ({
  type: ADD_CREW,
  ...entities
});

const getCrew = async (showId) => await axios.get(`${baseUrl}shows/${showId}/crew`);

export const handleCrewData = (data) => {
  const { entities, result } = normalize(data, crewSchema);
  const crewMembers = result.reduce((acc, crewMember) => {
    return {
      ...acc,
      [crewMember.person]: _.union(_.get(acc, crewMember.person, []), [ crewMember.type ])
    };
  }, {});
  return {
    ...entities,
    result,
    crewMembers
  };
};

export const requestCrew = (showId) => {
  return async (dispatch) => {
    const requestId = `GET_CREW_${showId}`;
    dispatch(loadingStart(requestId));
    try {
      const { data } = await getCrew(showId);
      const { person: people, crewMembers } = handleCrewData(data);
      dispatch(addCrew({ id: showId, people, crewMembers }));
      dispatch(loadingEnd(requestId));
    }
    catch(e) {
      dispatch(loadingEnd(requestId));
      throw e;
    }
  };
};
