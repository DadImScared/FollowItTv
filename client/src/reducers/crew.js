
import { ADD_CREW } from '../actiontypes/crew';

function crew(state = {}, action) {
  switch (action.type) {
    case ADD_CREW:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          ...action.crewMembers
        }
      };
    default:
      return state;
  }
}

export default crew;
