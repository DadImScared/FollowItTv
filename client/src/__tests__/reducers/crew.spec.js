
import { ADD_CREW } from '../../actiontypes/crew';
import reducer from '../../reducers/crew';

describe('crew reducer', () => {
  it('should add crew', () => {
    expect(
      reducer({}, {
        type: ADD_CREW,
        crewMembers: {
          1: ['role1', 'role2'],
          2: ['role3', 'role4']
        },
        id: 1
      })
    ).toMatchSnapshot();
  });
});
