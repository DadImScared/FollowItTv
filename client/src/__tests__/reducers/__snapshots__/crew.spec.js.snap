// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crew reducer should add crew 1`] = `
Object {
  "1": Object {
    "1": Array [
      "role1",
      "role2",
    ],
    "2": Array [
      "role3",
      "role4",
    ],
  },
}
`;
